### ISSUE TITLE
File write streams not flushed after writing undefined

### ISSUE THREAD

**Issue by psung** [Sage/ez-streams#41]
(Lifted from thread in streamline.js Google group)

The following program, which is supposed to write some lines to a file, typically exits without having written anything to the file (leaving a file of length 0).

```javascript
#!/usr/bin/env _node

var ez = require('ez-streams');

function main(_) {
  var outputFile = ez.devices.file.text.writer('testoutput.txt');
  var i;
  for (i = 0; i < 100; ++i) {
    outputFile.write(_, 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa', 'utf-8');
    outputFile.write(_, '\n', 'utf-8');
  }
  outputFile.write(_, undefined);
}

main(function (err) {
  if (err) {
    console.error(err.stack);
    process.exit(1);
  }
  process.exit();
});
```

This is on:
* streamline 1.0.13
* ez-streams 1.2.1
* streamline-runtime 1.0.16

I presume it's a race condition between the flushing of the stream data and the exit of the program, because if I add a `setTimeout(_, 1000)` after the call to end(), the data is written.

(In the past, under streamline 0.10.17 and ez-streams 0.1.7, what I did was to invoke end() with a callback, but the end method no longer takes a callback argument.)

**Comment by bjouhier** [comment]
@psung The fix was much simpler than I thought. I'm going to run more tests tomorrow and I'll publish to NPM if all goes well.

The repro was very helpful. Thanks.

**Comment by bjouhier** [comment]
The fix works on node 0.12 and 4.2 but the http unit test fails on 0.10 (travis-ci is so cool :+1: ). 

This is due to an inconsistency in the 0.10 APIs: the [`writable.end`](https://nodejs.org/docs/latest-v0.10.x/api/stream.html#stream_writable_end_chunk_encoding_callback) method accepts the callback argument  but the http [`response.end`](https://nodejs.org/docs/latest-v0.10.x/api/http.html#http_response_end_data_encoding) method rejects it. So the http `response` object is not a valid `writable` in node 0.10!

I'm getting this error on http `response` but it's likely that other node 0.10 streams will have similar issues. I don't want to invest much time into workarounds for node 0.10 streams. So I propose to bump the `ez-streams` version to 3.0 and put the node.js version floor at 0.12. People who want to use `ez-streams` with 0.10 can still get version 2.0 but they won't get this fix. @psung, would that work for you?

**Comment by psung** [comment]
Sounds good to me, thanks for the fix!

**Comment by bjouhier** [comment]
published to npm as 3.0.0 (flagged as beta)
